### PR TITLE
Setup tox and finalize frontend

### PR DESCRIPTION
**Pull request by jhubaum**
This MR sets up a basic testing workflow using tox and adds a github action to execute the testing pipeline whenever a branch is pushed to a pull request. It also fixes the visualization of the charging windows in the frontend. I will do refactoring of how the time data is stored in the database in a follow-up PR.

## Pre-merge checklist
- [x] Successfully deploy server and test the visualization of charging windows
  - This currently doesn't work since, for some reason, the celery worker is unable to connect to the database
- [x] Squash commits

### CODE REVIEW

**Comment by jhubaum** [comment]
@philippschmalen Can you change the protection settings of `main` so that we're only able to merge an MR when all checks have passed?

**Comment by philippschmalen** [comment]
Unfortunately, I cannot select the `Run Python CI` yet as a status check. I guess we first have to merge everything before I can specify it.

image

File: tests/test_config.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import yaml
```
Comment (philippschmalen) [inline tests/test_config.py:1]: Was confused first if you meant to create `conftest.py` like https://stackoverflow.com/questions/34466027/in-pytest-what-is-the-use-of-conftest-py-files , but it's used for fixtures.
Then I found test_config.py in the pytest repo https://github.com/pytest-dev/pytest/blob/main/testing/test_config.py . So, now I get it :)

File: src/data/redis_pandas_handler.py [hunk 2]
```diff
@@ -12,6 +13,18 @@ def datetimeindex_to_unix(df: pd.DataFrame) -> pd.DataFrame:
     return df
 
 
+def datetime_to_timestamp(dt: datetime.datetime):
+    return int(dt.timestamp()) // 10**9
+
+
+def unixtimeindex_to_datetime(df: pd.DataFrame) -> pd.DataFrame:
+    "Convert pandas unix timestamp index to datetime"
+    df = df.copy()
```
Comment (philippschmalen) [inline src/data/redis_pandas_handler.py:22]: Avoid weird issues with `df = df.copy()` , different naming instead: `df_cp = df.copy()`

File: src/data/redis_pandas_handler.py [hunk 3]
```diff
@@ -52,17 +65,48 @@ def dataframe_to_timeseries(
...
+        end_time: Optional[datetime.datetime] = None,
+        dtype=None,
+        name: Optional[str] = None,
+        index_name: Optional[str] = None,
+    ) -> pd.Series:
+        try:
+            timeseries = redis.ts().range(
+                redis_key,
+                "-" if start_time is None else datetime_to_timestamp(start_time),
+                "+" if end_time is None else datetime_to_timestamp(end_time),
```
Comment (philippschmalen) [inline src/data/redis_pandas_handler.py:82]: Nice!

**Review by philippschmalen — CHANGES_REQUESTED** [review]
Nice work! Just some details to discuss.

File: src/backend/api.py [hunk 4]
```diff
@@ -31,8 +37,74 @@ def disable(self):
     def toggle(self):
         self._publish("toggle")
 
+    def redis_timeseries_key(self, name):
+        return f"ts::device::{self.device.id}::{name}"
```
[the remark this answers is not in the document]
Comment (jhubaum) [reply src/backend/api.py]: I agree, and want to refactor most of the current redis interaction. I renamed the device keys in 4f11fd9 and would refactor the interaction in a follow-up PR. Or do you want me to do this before merging?

File: tests/test_config.py (hunk 1 above)
Comment (jhubaum) [reply tests/test_config.py:1]: So, nothing to do here, right? Or do you want to me to rename the test file?

File: src/data/redis_pandas_handler.py (hunk 2 above)
Comment (jhubaum) [reply src/data/redis_pandas_handler.py:22]: What weird issues? FWIW, `datetimeindex_to_unix` does the same thing and I want to refactor the redis handling in a way where both of these functions will become obsolete anyway. Unless you strongly object, I will ignore this for this MR.

**Comment by jhubaum** [comment]
From my side this is ready to merge. @philippschmalen Do you have any other objections?

**Review by philippschmalen — APPROVED** [review]

**Comment by philippschmalen** [comment]
> Before, we only stored the charging state in Redis. This is enough to
> control devices using MQTT. To visualize the charging windows, however,
> we also need the predicted value from the optimizer.

Instead of writing "a wrapper function to split a dataframe into multiple timeseries" we could query Redis twice for `ts:preprocessed` (values used for prediction) and `ts:charge`. So, we have the choice between 
1. Query Redis 2x on ts:charge, ts:preprocessed and merge on timestamp
2. write a wrapper that splits a dataframe

I prefer (1) as it introduces a dynamic mapping btw. a dataframe and Redis. Better to keep it deterministic to be more transparent & robust.
In addition to that, we need to refactor `src.data.visualize_intervals:plot_charging_windows` in a way that it handles unix timestamps within the dataframe and converts those to human-readable ones in user timezone.  Plotly handles the overlay of the two timeseries automatically, when they have the same length, throwing exception otherwise.

**Comment by jhubaum** [comment]
> Instead of writing "a wrapper function to split a dataframe into multiple timeseries" we could query Redis twice for ts:preprocessed (values used for prediction) and ts:charge.

Good point, I hadn't considered that. I agree, (1) is the better option. Since this MR was supposed to be only about fixing the frontend, I will implement this in my follow-up refactoring MR. I want to merge this now, so that we have a first running version available on `main` before I start breaking things again :D

I'll look into refactoring `plot_charging_windows` there as well.

**Pull request merged**